py-rse/bash-basics.Rmd

```diff
@@ -1847,65 +1810,69 @@ $ for filename in s*.txt; do echo $filename; head -n 19 $filename | tail -n 11;
 
 ## How can I create new filenames automatically? {#py-rse-bash-basics-autoname}
 
-Suppose we want to create a backup copy of each book whose name has the preposition 'of' in it.
+Suppose we want to create a backup copy of each book whose name ends in "e".
 If we don't want to change the files' names,
 we can do this with `cp`:
 
 ```shell
 $ cd ~/zipfs-law
 $ mkdir backup
-$ cp data/*of*.txt backup
+$ cp data/*e.txt backup
+$ ls backup
 ```
 
-But what if we want the backup files to have the [extension][filename-extension] `.bak` instead of `.txt`?
+```text
+jane_eyre.txt  time_machine.txt
+```
+
+But what if we want the backup files to have the [extension][filename-extension] `.bak` appended?
 `cp` can do this for a single file:
 
 ```shell
-$ cp data/anne_of_green_gables.txt backup/anne_of_green_gables.bak
+$ cp data/time_machine.txt backup/time_machine.txt.bak
 ```
 
 but can't do all the files at once:
 
 ```shell
-$ cp data/*of*.txt backup/*of*.bak
+$ cp data/*e.txt backup/*e.txt.bak
 ```
 
 ```text
-usage: cp [-R [-H | -L | -P]] [-fi | -n] [-apvXc] source_file target_file
-       cp [-R [-H | -L | -P]] [-fi | -n] [-apvXc] source_file ... target_directory
+cp: target 'backup/*e.txt.bak' is not a directory
 ```
 
-`backup/*of*.bak` doesn't match anything—those files don't yet exist—so
+`backup/*e.txt.bak` doesn't match anything—those files don't yet exist—so
+after the shell expands the wildcards,
 what we are actually asking `cp` to do is:
 
 ```shell
-$ cp data/anne_of_green_gables.txt   data/life_of_frederick_douglass.txt
-data/count_of_monte_cristo.txt
+$ cp data/jane_eyre.txt data/time_machine.txt backup/*e.bak
 ```
 
 This doesn't work because `cp` only understands how to do two things:
 copy a single file to create another file,
 or copy a bunch of files into a directory.
 If we give it more than two names as arguments,
 it expects the last one to be a directory;
-since it isn't,
-it reports an error.
+since `backup/*e.bak` is not,
+`cp` reports an error.
 
 Instead,
-let's copy all the files and then rename them:
+let's use a loop to copy files to the backup directory
+and append the `.bak` suffix:
 
 ```shell
```

Comment:
I could be convinced either way here. The reason I changed this loop was that I have a slight aversion to teach copying twice since this is inefficient and not a habit learners should pick up. Ideally, we would no need `cd`, but I can't think of a way to remove it that does not contain substring matching to exclude the path prefix, which I think is beyond what we teach here.